File: examples/one_max.rs
//! OneMax: find the bit string with the most ones.
//!
//! The "hello world" of genetic algorithms: a binary genome, tournament selection, uniform
//! crossover and bit-flip mutation, with statistics per generation.
//!
//! ```text
//! cargo run --release --example one_max
//! ```

use genoxide::prelude::*;

const LEN: usize = 500;

fn main() -> Result<()> {
    let ga = Ga::builder(Binary::new(LEN)?)
        .population_size(100)
        .select(Tournament::new(3)?)
        .crossover(UniformCrossover::new())
        .mutate(BitFlip::per_gene(1.0 / LEN as f64)?)
        .seed(42)
        .build()?;

    let mut statistics = Statistics::new();
    let outcome = Engine::new(ga, |genome: &Bits| genome.count_ones() as f64)
        .stop_when(Stop::target(LEN as f64).or(Stop::generations(10_000)))
        .observe(&mut statistics)
        .run()?;

    println!("generation  best  mean");
    for record in statistics.records().iter().step_by(25) {
        println!(
            "{:>10}  {:>4}  {:>6.1}",
            record.generation,
            record.best.and_then(Fitness::score).unwrap_or(f64::NAN),
            record.mean.unwrap_or(f64::NAN),
        );
    }
    println!(
        "\n{:?} after {} generations and {} evaluations ({:?}): {}",
        outcome.stop_reason(),
        outcome.generations(),
        outcome.evaluations(),
        outcome.elapsed(),
        outcome.best_fitness(),
    );
    Ok(())
}

File: python/examples/onemax.py
"""OneMax, as in DEAP's first tutorial: the 100-bit string with the most ones."""

import genoxide as gx

ga = gx.Ga(
    gx.Binary(100),
    population_size=300,
    select=gx.Tournament(3),
    crossover=gx.PointCrossover(2),
    crossover_rate=0.5,
    mutation=gx.BitFlip(rate=0.05),
    mutation_rate=0.2,
    seed=1,
)
result = ga.run(lambda bits: bits.sum(), target=100, generations=1_000)
print(f"{result.best_fitness:.0f} ones after {result.generations} generations ({result.stop_reason})")
